src/pr-trigger.ts

```diff
@@ -4,175 +4,147 @@ import { queryPRInfo, deriveStateForPR } from "./pr-info";
-    }
-
-    const webhook = req.body;
-    const action = "action" in webhook ? webhook.action : "status";
-
-    if (webhook.sender.login === "typescript-bot") {
-        context.log.info(`Skipped webhook because it was triggered by typescript-bot`);
-        context.res = {
-            status: 204,
-            body: `NOOPing because typescript-bot triggered the request`
-        };
-        return;
-    }
-
-    if (!allowListedActions.includes(action) && !allowListedActions.includes("*")) {
-        context.log.info(`Skipped webhook, ${action} on ${event}, do not know how to handle the action`);
-        context.res = {
-            status: 204,
-            body: `NOOPing due to not supporting ${action} on ${event}`
-        };
-        return;
-    }
-
-    let prNumber = -1;
-    let prTitle = "(title not fetched)"; // this is only used for logging, not worth an API lookup => not always set
-    if ("pull_request" in webhook) {
-        prNumber = webhook.pull_request.number;
-        prTitle = webhook.pull_request.title;
-    } else if ("issue" in webhook) {
-        prNumber = webhook.issue.number;
-        prTitle = webhook.issue.title;
-    } else if ("project_card" in webhook) {
-        // Hack to get the PR number, could be found directly only in `content_url`
-        const cardURL = webhook.project_card.content_url;
-        const numberInURL = cardURL.match(/\/\d+$/);
-        if (!numberInURL) throw new Error(`Could not get PR for project card URL: ${cardURL}`);
-        prNumber = +numberInURL[0].substring(1);
-    } else if ("check_suite" in webhook) {
-        // See https://github.com/maintainers/early-access-feedback/issues/114 for more context on getting a PR from a SHA
-        // TLDR: it's not in the API, and this search hack has been in used on Peril for the last ~3 years
-        // => Instead of an arbitrary search, use `associatedPullRequests` (https://developer.github.com/v4/changelog/2019-03-08-schema-changes/)
-        const owner = webhook.repository.owner.login;
-        const repo = webhook.repository.name;
-        const sha = webhook.check_suite.head_sha;
-        const pr = await runQueryToGetPRMetadataForSHA1(owner, repo, sha);
-        if (!pr || pr.closed) {
-            const whatFailed = !pr ? "a PR" : "an open PR";
-            context.log.info(`Skipped webhook, could not find ${whatFailed} for the sha referenced in the status (${sha})`);
+    const eventHandler = createEventHandler({});
+    eventHandler.on([
+        "check_suite.completed",
+        "issue_comment.created",
+        "issue_comment.deleted",
+        "issue_comment.edited",
+        "project_card.moved",
+        "pull_request.closed",
+        "pull_request.edited",
+        "pull_request.opened",
+        "pull_request.ready_for_review",
+        "pull_request.reopened",
+        "pull_request.synchronize",
+        "pull_request_review.dismissed",
+        "pull_request_review.submitted",
+    ], async event => {
```

Comment:
Done, I moved it to a function. :heavy_check_mark:

Reply:
But that's not a toplevel function...  Since you did the hard part of fighting with the types, I can write what I'm suggesting:

```ts
export async function httpTrigger(context: Context, req: HttpRequest) {
    ...
    eventHandler.on(eventNames as unknown as typeof eventNames[number], handleTrigger(context));
    ...
}

const handleTrigger = (context: Context) => async (event: EmitterWebhookEvent<typeof eventNames[number]>) => {
    ...
};
```

and even better, move this too:

```ts
const prFromEvent = (event: EmitterWebhookEvent<typeof eventNames[number]>) => {
    switch (event.name) ...
};
```